File: performance/load-test.js
import http from 'k6/http';
import { check, sleep } from 'k6';
import { Trend, Rate, Counter } from 'k6/metrics';

// Custom metric trends for per-endpoint breakdown
const ordersTrend = new Trend('orders_duration');
const inventoryTrend = new Trend('inventory_duration');
const machinesTrend = new Trend('machines_duration');
const orderCreationTrend = new Trend('order_creation_duration');
const quoteRequestTrend = new Trend('quote_request_duration');
const paymentsTrend = new Trend('payments_duration');
const errorRate = new Rate('error_rate');
const successfulRequests = new Counter('successful_requests');

export const options = {
  scenarios: {
    manufacturing_workload: {
      executor: 'ramping-vus',
      startVUs: 0,
      stages: [
        { duration: __ENV.RAMP_UP || '5m', target: 100 },    // Ramp-up: 0 -> 100 users
        { duration: __ENV.HOLD || '10m', target: 100 },      // Steady hold: 100 users
        { duration: __ENV.RAMP_DOWN || '2m', target: 0 },    // Ramp-down: 100 -> 0 users
      ],
      gracefulRampDown: '30s',
    },
  },
  thresholds: {
    // SLA 1: p95 latency < 1000ms
    http_req_duration: ['p(95)<1000', 'p(99)<2000'],
    // SLA 2: Error rate < 1%
    error_rate: ['rate<0.01'],
    http_req_failed: ['rate<0.01'],
    // SLA 3: Throughput >= 100 req/sec
    http_reqs: ['rate>=100'],
  },
};

const BASE_URL = __ENV.BASE_URL || 'http://localhost:3000';

const JSON_HEADERS = {
  'Content-Type': 'application/json',
  'Accept': 'application/json',
  'X-Service-Key': 'forgeiq_internal_service_key_2026',
  'X-Org-ID': 'factory_acme_01',
};

export default function () {
  const roll = Math.random() * 100;

  // 1. 30% Workload: GET /orders
  if (roll < 30) {
    const res = http.get(`${BASE_URL}/api/orders`, { headers: JSON_HEADERS, timeout: '5s' });
    ordersTrend.add(res.timings.duration);
    const pass = check(res, {
      'GET /orders status 200': (r) => r.status === 200,
      'GET /orders latency < 1000ms': (r) => r.timings.duration < 1000,
    });
    errorRate.add(!pass);
    if (pass) successfulRequests.add(1);
  }
  // 2. 20% Workload: GET /inventory
  else if (roll < 50) {
    const res = http.get(`${BASE_URL}/api/inventory`, { headers: JSON_HEADERS, timeout: '5s' });
    inventoryTrend.add(res.timings.duration);
    const pass = check(res, {
      'GET /inventory status 200': (r) => r.status === 200,
      'GET /inventory latency < 1000ms': (r) => r.timings.duration < 1000,
    });
    errorRate.add(!pass);
    if (pass) successfulRequests.add(1);
  }
  // 3. 15% Workload: GET /machines
  else if (roll < 65) {
    const res = http.get(`${BASE_URL}/api/machines`, { headers: JSON_HEADERS, timeout: '5s' });
    machinesTrend.add(res.timings.duration);
    const pass = check(res, {
      'GET /machines status 200': (r) => r.status === 200,
      'GET /machines latency < 1000ms': (r) => r.timings.duration < 1000,
    });
    errorRate.add(!pass);
    if (pass) successfulRequests.add(1);
  }
  // 4. 20% Workload: POST /orders (Create active order)
  else if (roll < 85) {
    const payload = JSON.stringify({
      title: `Laser Cut Flange Batch #${Math.floor(Math.random() * 10000)}`,
      customerName: 'Apex Aerospace Solutions',
      priority: 'Normal',
      totalAmount: 38500,
      dueDate: '2026-09-30',
      materialSku: 'RAW-SS304-18G',
      quantityUnits: 50,
    });
    const res = http.post(`${BASE_URL}/api/orders`, payload, { headers: JSON_HEADERS, timeout: '5s' });
    orderCreationTrend.add(res.timings.duration);
    const pass = check(res, {
      'POST /orders status 200': (r) => r.status === 200 || r.status === 201,
      'POST /orders latency < 1000ms': (r) => r.timings.duration < 1000,
    });
    errorRate.add(!pass);
    if (pass) successfulRequests.add(1);
  }
  // 5. 10% Workload: POST /ai/quote-request (Submit RFQ / AI Estimation)
  else if (roll < 95) {
    const payload = JSON.stringify({
      material: 'SS304',
      thickness: 3.0,
      cutLengthMm: 1800.0,
      pierceCount: 8,
      bendCount: 4,
      quantity: 100,
    });
    const res = http.post(`${BASE_URL}/api/ai/quote-request`, payload, { headers: JSON_HEADERS, timeout: '5s' });
    quoteRequestTrend.add(res.timings.duration);
    const pass = check(res, {
      'POST /ai/quote-request status 200': (r) => r.status === 200,
      'POST /ai/quote-request latency < 1500ms': (r) => r.timings.duration < 1500,
    });
    errorRate.add(!pass);
    if (pass) successfulRequests.add(1);
  }
  // 6. 5% Workload: POST /payments (Create payment checkout)
  else {
    const payload = JSON.stringify({
      amount: 14500,
      currency: 'INR',
      receipt: `rcpt_load_${Date.now()}`,
    });
    const res = http.post(`${BASE_URL}/api/payments`, payload, { headers: JSON_HEADERS, timeout: '5s' });
    paymentsTrend.add(res.timings.duration);
    const pass = check(res, {
      'POST /payments status 200': (r) => r.status === 200,
      'POST /payments latency < 1000ms': (r) => r.timings.duration < 1000,
    });
    errorRate.add(!pass);
    if (pass) successfulRequests.add(1);
  }

  // Realistic thinking time between operations (200ms - 800ms)
  sleep(Math.random() * 0.6 + 0.2);
}

// Summary Report Generation
export function handleSummary(data) {
  const p95 = data.metrics.http_req_duration ? data.metrics.http_req_duration.values['p(95)'].toFixed(2) : 'N/A';
  const p99 = data.metrics.http_req_duration ? data.metrics.http_req_duration.values['p(99)'].toFixed(2) : 'N/A';
  const mean = data.metrics.http_req_duration ? data.metrics.http_req_duration.values.avg.toFixed(2) : 'N/A';
  const totalReqs = data.metrics.http_reqs ? data.metrics.http_reqs.values.count : 0;
  const rate = data.metrics.http_reqs ? data.metrics.http_reqs.values.rate.toFixed(1) : 'N/A';
  const errPct = data.metrics.http_req_failed ? (data.metrics.http_req_failed.values.rate * 100).toFixed(2) : '0.00';

  console.log('===============================================================');
  console.log('FORGEIQ K6 HIGH-CONCURRENCY LOAD TEST SUMMARY');
  console.log('===============================================================');
  console.log(`  Total Requests    : ${totalReqs}`);
  console.log(`  Throughput Rate   : ${rate} req/sec`);
  console.log(`  Mean Latency      : ${mean} ms`);
  console.log(`  p95 Latency       : ${p95} ms (SLA: < 1000 ms)`);
  console.log(`  p99 Latency       : ${p99} ms (SLA: < 2000 ms)`);
  console.log(`  Error Rate        : ${errPct} % (SLA: < 1.0 %)`);
  console.log('===============================================================');

  return {
    'performance/results/k6_summary.json': JSON.stringify(data, null, 2),
  };
}

File: performance/load_test.py
"""
ForgeIQ Production Load Test Suite - Locust Framework
Target: 100 Concurrent Users, 17-Minute Ramp/Hold/Ramp-down Soak Test
Workload Distribution:
  - 30%: GET /api/orders
  - 20%: GET /api/inventory
  - 15%: GET /api/machines
  - 20%: POST /api/orders
  - 10%: POST /ai/quote-request
  - 5%: POST /payments
"""

import json
import random
import time
from locust import HttpUser, task, between, LoadTestShape, events

HEADERS = {
    "Content-Type": "application/json",
    "Accept": "application/json",
    "X-Service-Key": "forgeiq_internal_service_key_2026",
    "X-Org-ID": "factory_acme_01",
}

class ForgeIQUser(HttpUser):
    # Simulated realistic user think time: 0.2s - 0.8s
    wait_time = between(0.2, 0.8)

    # 1. 30% Workload: GET /orders (read recent orders)
    @task(30)
    def get_orders(self):
        with self.client.get("/api/orders", headers=HEADERS, name="GET /orders", catch_response=True) as response:
            if response.status_code == 200:
                if response.elapsed.total_seconds() > 5.0:
                    response.failure(f"Timeout SLA breached (>5000ms): {response.elapsed.total_seconds()}s")
                else:
                    response.success()
            else:
                response.failure(f"Unexpected status: {response.status_code}")

    # 2. 20% Workload: GET /inventory (check stock)
    @task(20)
    def get_inventory(self):
        with self.client.get("/api/inventory", headers=HEADERS, name="GET /inventory", catch_response=True) as response:
            if response.status_code == 200:
                if response.elapsed.total_seconds() > 5.0:
                    response.failure(f"Timeout SLA breached (>5000ms): {response.elapsed.total_seconds()}s")
                else:
                    response.success()
            else:
                response.failure(f"Unexpected status: {response.status_code}")

    # 3. 15% Workload: GET /machines (view equipment)
    @task(15)
    def get_machines(self):
        with self.client.get("/api/machines", headers=HEADERS, name="GET /machines", catch_response=True) as response:
            if response.status_code == 200:
                if response.elapsed.total_seconds() > 5.0:
                    response.failure(f"Timeout SLA breached (>5000ms): {response.elapsed.total_seconds()}s")
                else:
                    response.success()
            else:
                response.failure(f"Unexpected status: {response.status_code}")

    # 4. 20% Workload: POST /orders (create new order)
    @task(20)
    def post_orders(self):
        order_payload = {
            "title": f"Precision Bracket Run #{random.randint(1000, 99999)}",
            "customerName": "Titan Industrial Dynamics",
            "priority": "Normal",
            "totalAmount": random.choice([25000, 48000, 75000]),
            "dueDate": "2026-10-15",
            "materialSku": "RAW-SS304-18G",
            "quantityUnits": random.randint(10, 200)
        }
        with self.client.post("/api/orders", json=order_payload, headers=HEADERS, name="POST /orders", catch_response=True) as response:
            if response.status_code in (200, 201):
                if response.elapsed.total_seconds() > 5.0:
                    response.failure(f"Timeout SLA breached (>5000ms): {response.elapsed.total_seconds()}s")
                else:
                    response.success()
            else:
                response.failure(f"Order creation failed: {response.status_code}")

    # 5. 10% Workload: POST /ai/quote-request (submit RFQ / AI estimation)
    @task(10)
    def post_ai_quote_request(self):
        rfq_payload = {
            "material": random.choice(["SS304", "AL6061", "MildSteel"]),
            "thickness": random.choice([1.5, 3.0, 5.0]),
            "cutLengthMm": round(random.uniform(500.0, 3200.0), 1),
            "pierceCount": random.randint(4, 30),
            "bendCount": random.randint(0, 8),
            "quantity": random.randint(50, 500)
        }
        with self.client.post("/api/ai/quote-request", json=rfq_payload, headers=HEADERS, name="POST /ai/quote-request", catch_response=True) as response:
            if response.status_code == 200:
                if response.elapsed.total_seconds() > 5.0:
                    response.failure(f"Timeout SLA breached (>5000ms): {response.elapsed.total_seconds()}s")
                else:
                    response.success()
            else:
                response.failure(f"AI Quote request failed: {response.status_code}")

    # 6. 5% Workload: POST /payments (create payment checkout)
    @task(5)
    def post_payments(self):
        payment_payload = {
            "amount": random.choice([12000, 35000, 68000]),
            "currency": "INR",
            "receipt": f"rcpt_locust_{int(time.time() * 1000)}"
        }
        with self.client.post("/api/payments", json=payment_payload, headers=HEADERS, name="POST /payments", catch_response=True) as response:
            if response.status_code == 200:
                if response.elapsed.total_seconds() > 5.0:
                    response.failure(f"Timeout SLA breached (>5000ms): {response.elapsed.total_seconds()}s")
                else:
                    response.success()
            else:
                response.failure(f"Payment initiation failed: {response.status_code}")


class ManufacturingSoakShape(LoadTestShape):
    """
    Automated 17-minute stage profile:
    - Stage 1: Ramp up 0 -> 100 users over 5 minutes (300 seconds)
    - Stage 2: Hold 100 users for 10 minutes (600 seconds -> total 900s)
    - Stage 3: Ramp down 100 -> 0 users over 2 minutes (120 seconds -> total 1020s)
    """
    stages = [
        {"duration": 300, "users": 100, "spawn_rate": 0.33},   # 0 -> 100 users in 300s
        {"duration": 900, "users": 100, "spawn_rate": 10},     # Hold 100 users for 600s
        {"duration": 1020, "users": 0, "spawn_rate": 0.83},    # 100 -> 0 users in 120s
    ]

    def tick(self):
        run_time = self.get_run_time()
        for stage in self.stages:
            if run_time < stage["duration"]:
                tick_data = (stage["users"], stage["spawn_rate"])
                return tick_data
        return None
